src/set_theory/ordinal/basic.lean

```diff
@@ -358,28 +358,29 @@ injective_of_increasing r (<) (typein r) (λ x y, (typein_lt_typein r).2)
   {a b} : typein r a = typein r b ↔ a = b :=
 (typein_injective r).eq_iff
 
+/-- Principal segment version of the `typein` function, embedding a well order into
+  ordinals as a principal segment. -/
+def typein.principal_seg (r : α → α → Prop) [is_well_order α r] :
+  r ≺i ((<) : ordinal → ordinal → Prop) :=
+⟨⟨⟨typein r, typein_injective r⟩, λ a b, typein_lt_typein r⟩, type r,
+  λ o, ⟨typein_surj r, λ ⟨a, h⟩, h ▸ typein_lt_type r a⟩⟩
+
+@[simp] theorem typein.principal_seg_coe (r : α → α → Prop) [is_well_order α r] :
+  (typein.principal_seg r : α → ordinal) = typein r := rfl
+
 /-! ### Enumerating elements in a well-order with ordinals. -/
 
 /-- `enum r o h` is the `o`-th element of `α` ordered by `r`.
   That is, `enum` maps an initial segment of the ordinals, those
   less than the order type of `r`, to the elements of `α`. -/
-def enum (r : α → α → Prop) [is_well_order α r] (o) : o < type r → α :=
-quot.rec_on o (λ ⟨β, s, _⟩ h, (classical.choice h).top) $
-λ ⟨β, s, _⟩ ⟨γ, t, _⟩ ⟨h⟩, begin
-  resetI, refine funext (λ (H₂ : type t < type r), _),
-  have H₁ : type s < type r, {rwa type_eq.2 ⟨h⟩},
-  have : ∀ {o e} (H : o < type r), @@eq.rec
-   (λ (o : ordinal), o < type r → α)
-   (λ (h : type s < type r), (classical.choice h).top)
-     e H = (classical.choice H₁).top, {intros, subst e},
-  exact (this H₂).trans (principal_seg.top_eq h
-    (classical.choice H₁) (classical.choice H₂))
-end
+def enum (r : α → α → Prop) [is_well_order α r] (o) (h : o < type r) : α :=
+(typein.principal_seg r).iso_subrel.symm ⟨o, h⟩
```

Comment:
Is there any point having both `ordinal.enum` and `ordinal.enum_iso` now that they are both one-liners from your new definition?

Reply:
`enum` should definitely stay, given its heavy use throughout the ordinal files. `enum_iso` should probably stay too, since it's a natural theorem statement with a nontrivial proof (even if a one-liner).

Reply:
Yeah but what I mean is that `enum` is just one direction of `enum_iso`, so we could replace `enum` by `enum_iso`, nothing would break and we would have one less definition. (and keep the name `enum` because there's no need to specify the amount of bundling anymore)

Reply:
I don't think it's fair calling it "just one direction", since `(typein.principal_seg r).iso_subrel.symm ⟨o, h⟩` is definitely a mouthful compared to `enum r o h`.

Reply:
You're misunderstanding me. With my suggestion, `enum r o h` would become `enum r ⟨o, h⟩` (and Lean will insert a `coe_fn`).

Reply:
Ah, I do agree with that then.

Reply:
https://github.com/leanprover-community/mathlib/pull/18695